examples/merkle_key.rs

```diff
@@ -0,0 +1,146 @@
+// Copyright 2020-2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+//! cargo run --example merkle_key
+
+use identity::core::BitSet;
+use identity::core::FromJson;
+use identity::core::Object;
+use identity::core::ToJson;
+use identity::core::Url;
+use identity::credential::CredentialBuilder;
+use identity::credential::Subject;
+use identity::credential::VerifiableCredential;
+use identity::crypto::merkle_tree::MTree;
+use identity::crypto::merkle_tree::Proof;
+use identity::crypto::KeyCollection;
+use identity::crypto::MerkleKey;
+use identity::crypto::MerkleKeySignerEd25519;
+use identity::crypto::MerkleKeyVerifierEd25519;
+use identity::crypto::PublicKey;
+use identity::crypto::SecretKey;
+use identity::crypto::SignatureOptions;
+use identity::did::verifiable::Properties;
+use identity::did::Document;
+use identity::did::DocumentBuilder;
+use identity::did::Method;
+use identity::did::MethodBuilder;
+use identity::did::MethodData;
+use identity::did::MethodType;
+use identity::did::DID;
+use rand::rngs::OsRng;
+use rand::Rng;
+use sha2::Sha256;
+
+type VerifiableDocument = Document<Properties, Object, ()>;
+
+type Signer<'a> = MerkleKeySignerEd25519<'a, Sha256>;
+type Verifier<'a> = MerkleKeyVerifierEd25519<'a, Sha256>;
+
+const LEAVES: usize = 1 << 10;
+
+fn main() {
+  // Generate a collection of ed25519 keys
+  let keys: KeyCollection = KeyCollection::new_ed25519(LEAVES).unwrap();
+
+  // Construct a Merkle tree from the public keys
+  let tree: MTree<Sha256> = keys.to_merkle_tree().unwrap();
+
+  // Generate a Merkle Key Collection public key value with ed25519 as the
+  // signature algorithm, SHA-256 as the digest algorithm, and the Merkle root
+  // of the key collection
+  let key_data: Vec<u8> = MerkleKey::encode_ed25519_key::<Sha256>(&tree);
+
+  // Create a DID for the document owner
+  let controller: DID = "did:example:123".parse().unwrap();
+
+  // Create a `MerkleKeyCollection2021` verification method with
+  // the previously generated public key value
+  let method: Method<Object> = MethodBuilder::default()
+    .id(controller.join("#key-collection").unwrap())
+    .controller(controller.clone())
+    .key_type(MethodType::MerkleKeyCollection2021)
+    .key_data(MethodData::new_b58(&key_data))
+    .build()
+    .unwrap();
+
+  println!("method: {:#}", method);
+
+  // Create a DID Document with the `MerkleKeyCollection2021` method
+  // as the sole verification method
```

Comment:
The example currently doesn't write to or read from the Tangle. While it does test all the appropriate functionalities, this is not how it would be utilized in the real world. I suggest we update the example to make use of the Tangle. This also means we can't add this as the "sole verification method", but would need to add a normal authentication-key for signing the DID Documents.

Reply:
👍 I'll add some more realistic Tangle usage to the example